### PR TITLE
Buildtools for compiling dart as a library

### PR DESCRIPTION
**Pull request by hkershaw-brown**
## Description:

Buildtools for compiling dart as a library (libdart.a or libdart.so)

SiParCS 2024 [Anh Pham](https://github.com/anhpham3621) and [Suman Shekhar](https://github.com/Sumanshekhar17) students used dart as a static library.  
There are other people interested in having dart as a library, so this provides buildtools so they don't have to create Makefiles by hand. 

This pull request has a developer test directory 'library' which has a shared lib build and a static lib build and a little test script to demo building and linking filter.

Not sure what level of docs to add, maybe just a section in [DART build system](https://docs.dart.ucar.edu/en/latest/guide/quickbuild.html) for people who want to play with a library.

### Fixes issue

no issue

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Documentation changes needed?

- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.

### Tests
DART/developer_tests/library/test_libraries.sh does a bitwise test for filter (regular build vs. static library vs shared library)
edit: also did a paranoia check that -fPIC was bitwise for a regular build of filter.

## Checklist for merging

- [ ] Updated changelog entry
- [ ] Documentation updated
- [ ] Update conf.py

## Checklist for release
- [ ] Merge into main
- [ ] Create release from the main branch with appropriate tag
- [ ] Delete feature-branch

## Testing Datasets

- [ ] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [ ] No dataset needed

### CODE REVIEW

File: developer_tests/library/mkmf.template.gfortran [hunk 1]
```diff
@@ -0,0 +1,147 @@
+# Template for GNU gfortran on Linux or Mac OSX
+#
+# DART software - Copyright UCAR. This open source software is provided
+# by UCAR, "as is", without charge, subject to all terms of use at
+# http://www.image.ucar.edu/DAReS/DART/DART_download
+#
+# DART $Id$
```
Comment (mjs2369) [inline developer_tests/library/mkmf.template.gfortran:7]:
```suggestion
```

Comment (mjs2369) [reply developer_tests/library/mkmf.template.gfortran:7]: Also need to remove the corresponding lines at the bottom of the file

Comment (hkershaw-brown) [reply developer_tests/library/mkmf.template.gfortran:7]: good point, I'll take out all this and the gfortran waffle in the comments. 
I just blindly copied the mkmf.template.gfortran

File: developer_tests/library/mkmf.template.gfortran [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+# typical use with mkmf
```
Comment (mjs2369) [inline developer_tests/library/mkmf.template.gfortran:9]: Do we need to have all of this additional text in the comments from here (line 9) until line 131? In the newer mkmf.templates we've released (mkmf.template.quickscat.gfortran for example) we don't include a lot of this info. Just 114-131.

Comment (hkershaw-brown) [reply developer_tests/library/mkmf.template.gfortran]: no I don't think so, (just took it out). I think it is better to keep it simple.

**Comment by mjs2369** [comment]
@hkershaw-brown I'm getting some errors when running test_libraries.sh

```
/glade/derecho/scratch/masmith/DART/assimilation_code/modules/utilities/mpi_utilities_mod.f90(1862): error #6404: This name does not have a type, and must have an explicit type.   [SYSTEM]
    rc = system(trim(shell_name)//' '//trim(execute)//' '//char(0))
---------^
/glade/derecho/scratch/masmith/DART/assimilation_code/modules/utilities/mpi_utilities_mod.f90(2033): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [MPI_UTILITIES_MOD]
 use mpi_utilities_mod, only : get_dart_mpi_comm
-----^
```

I'll keep looking into these, but I thought I'd let you know now to see if you might already know the causes.

**Comment by hkershaw-brown** [comment]
@mjs2369 are you running with a gfortran mkmf.template but he intel module on Derecho?

**Comment by mjs2369** [comment]
> @mjs2369 are you running with a gfortran mkmf.template but the intel module on Derecho?

Yep, duh. Though I did have to add the mpif08 argument to lines that run quickbuild (22, 36, and 44) in test_libraries.sh to get it compile further with gfortran. But it is still erroring out while trying to run filter for the first time on line 62 (the no lib build):
```
ar: creating libdart.a
a - algorithm_info_mod.o
a - bnrh_distribution_mod.o
a - netcdf_utilities_mod.o
a - schedule_mod.o
a - io_filenames_mod.o
.... (a bunch more files)
a - direct_netcdf_mod.o
a - distribution_params_mod.o
a - assert_mod.o
No host list provided
```

I haven't been able to figure out why this is happening yet

**Comment by mjs2369** [comment]
Some last questions/comments other than the error are:

I think minimal docs in the like you mentioned in the body of the PR sound like the way to go, but I think they might make more sense in [Compiling DART](https://docs.dart.ucar.edu/en/latest/guide/compiling-dart.html) instead of "DART Build System". 

We aren't supporting compiling DART as a library for Windows, right? The mkmf.template.gfortran is only for osx and linux, and I believe that Windows uses different file extensions for library files (static libraries are .lib in Windows instead of .a)

**Comment by hkershaw-brown** [comment]
> 
> ```
> ar: creating libdart.a
> a - algorithm_info_mod.o
> a - bnrh_distribution_mod.o
> a - netcdf_utilities_mod.o
> a - schedule_mod.o
> a - io_filenames_mod.o
> .... (a bunch more files)
> a - direct_netcdf_mod.o
> a - distribution_params_mod.o
> a - assert_mod.o
> No host list provided
> ```
> 
> I haven't been able to figure out why this is happening yet

Are you trying to run an mpi job on the Derecho login nodes.

**Comment by hkershaw-brown** [comment]
> Some last questions/comments other than the error are:
> 
> 1. I think minimal docs in the like you mentioned in the body of the PR sound like the way to go, but I think they might make more sense in [Compiling DART](https://docs.dart.ucar.edu/en/latest/guide/compiling-dart.html) instead of "DART Build System".

So this isn't a feature for general use.  It would be for people developing tools that use DART. I think the build system docs (developers section) is the place for it. 

> 2. We aren't supporting compiling DART as a library for Windows, right? The mkmf.template.gfortran is only for osx and linux, and I believe that Windows uses different file extensions for library files (static libraries are .lib in Windows instead of .a)

No we're not supporting DART as a library for windows. If people want to use DART on windows, then I'd recommend using WSL.

**Review by mjs2369 — APPROVED** [review]
Everything looks good; the library builds and developer test to check that they are bitwise work well. Approved

**Pull request merged**